# TASK 3 Import your analysis libraries, as well as your new customer data set as a dataframe.

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
# Set path 
path = r'C:\Users\peter\Documents\04-2024 Instacart Basket Analysis'

In [3]:
# Import new customer data set from the 'Original Data' folder as a dataframe
customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

In [4]:
# Preview the new dataframe
customers.head(10)

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5   133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6   152052      Chris    Walton    Male         Montana   20    1/1/2017   
7   168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8    69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9    82820      Shawn     Chung    Male        Virginia   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374  
5             2    married   49643  
6             0     single   61746  
7             0     single   63712  
8             1    married  162432  
9             2    married   32072

In [5]:
# Check the dimensions of the dataframe
customers.shape

(206209, 10)

In [6]:
# create and view descriptive statistics 
customers.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [7]:
# Check data types
customers.dtypes

user_id          int64
First Name      object
Surnam          object
Gender          object
STATE           object
Age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

In [8]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


# TASK 4 Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.

In [9]:
# Rename the columns for similar format
customers.rename(columns = {'First Name' : 'first_name'}, inplace = True)

In [10]:
customers.rename(columns = {'Surnam' : 'surname'}, inplace = True)

In [11]:
customers.rename(columns = {'Gender' : 'gender'}, inplace = True)

In [12]:
customers.rename(columns = {'Age' : 'age'}, inplace = True)

In [13]:
customers.rename(columns = {'STATE' : 'state'}, inplace = True)

In [14]:
customers.rename(columns = {'fam_status' : 'marital_status'}, inplace = True)

In [15]:
# Check columns
customers.columns

Index(['user_id', 'first_name', 'surname', 'gender', 'state', 'age',
       'date_joined', 'n_dependants', 'marital_status', 'income'],
      dtype='object')

In [16]:
customers.head(5)

user_id first_name   surname  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants marital_status  income  
0             3        married  165665  
1             0         single   59285  
2             2        married   99568  
3             0         single   42049  
4             1        married   40374

In [16]:
# Columns are succesfully renamed. Now they have a similar form
# I think analysis for each columns are possible / no columns will be dropped

# TASK 5 Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data.

In [17]:
# Checking missing values
customers.isnull().sum()

user_id               0
first_name        11259
surname               0
gender                0
state                 0
age                   0
date_joined           0
n_dependants          0
marital_status        0
income                0
dtype: int64

First name has 11259 missing values. As this is nota key variable for the analysis I do not need to address them.


In [18]:
# Checking duplicates
df_dups = customers[customers.duplicated()]

In [19]:
df_dups

Empty DataFrame
Columns: [user_id, first_name, surname, gender, state, age, date_joined, n_dependants, marital_status, income]
Index: []

no duplicates found

In [20]:
# Checking mixed-type
customers.dtypes

user_id            int64
first_name        object
surname           object
gender            object
state             object
age                int64
date_joined       object
n_dependants       int64
marital_status    object
income             int64
dtype: object

In [21]:
for col in customers.columns.tolist():
  weird = (customers[[col]].map(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customers[weird]) > 0:
    print (col)

first_name


first name has mixed type of data

In [22]:
# changing data type to first name to string
customers['first_name'] = customers['first_name'].astype('str')

In [23]:
# double check the data type
customers.dtypes

user_id            int64
first_name        object
surname           object
gender            object
state             object
age                int64
date_joined       object
n_dependants       int64
marital_status    object
income             int64
dtype: object

In [24]:
# Let's check again the mixed data-type
for col in customers.columns.tolist():
  weird = (customers[[col]].map(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customers[weird]) > 0:
    print (col)

no mixed data type cloumns and all columns has the right data type

# TASK 6 Combine your customer data with the rest of your prepared Instacart data. (Hint: Make sure the key columns are the same data type!)

In [25]:
# Import the prepared Instacart data
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined_merged_new_variables_derived_aggregated_visual.pkl'))

In [26]:
# Checking data type
ords_prods_merge.dtypes

order_id                     int64
user_id                      int64
order_number                 int64
orders_day_of_week           int64
order_hour_of_day            int64
days_since_prior_order     float64
product_id                   int64
add_to_cart_order            int64
reordered                    int64
product_name                object
aisle_id                     int64
department_id                int64
prices                     float64
_merge                    category
price_range_loc             object
busiest_day                 object
busiest_days                object
busiest_period_of_day       object
max_order                    int64
loyalty_flag                object
average_spend              float64
spender_flag                object
Customer_frequency         float64
frequency_flag              object
dtype: object

The key columns are user_id, they have the same data type int64

In [27]:
# Drop the '_merge' column from the previous exercise
ords_prods_merge = ords_prods_merge.drop(columns=['_merge'])

In [28]:
# Merge ords_prods_merge with customers data
ords_prods_customers_merge = ords_prods_merge.merge(customers, on = 'user_id', indicator = True)

In [29]:
ords_prods_customers_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     7.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  first_name  surname  gender    state age date_joined  \
0         Soda  ...       Linda   Nguyen  Female  Alabama  31   2/17/2019   
1         Soda  ...       Linda   Nguyen  Female  Alabama  31   2/17/2019   
2         Soda  ...       Linda   Nguyen  Female  Alabama  31   2/17/2019   
3         Soda  ...       Linda   Nguyen  Female  Alabama  31   2/17/2019   
4         Soda  ...       Linda   Nguyen  Female  Alabama  31   2/17/2019   

  n_dependants  marital_status income  _merge  
0            3         married  40423    both  
1            3         married  40423    both  
2            3         married  40423    both  
3            3         married  40423    both  
4            3         married  40423    both  

[5 rows x 33 columns]

# TASK 8 Export this new dataframe as a pickle file so you can continue to use it in the second part of this task.

In [30]:
# Export this new dataframe as a pickle file
ords_prods_customers_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_customers_aggregated.pkl'))

In [31]:
ords_prods_customers_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     7.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  first_name  surname  gender    state age date_joined  \
0         Soda  ...       Linda   Nguyen  Female  Alabama  31   2/17/2019   
1         Soda  ...       Linda   Nguyen  Female  Alabama  31   2/17/2019   
2         Soda  ...       Linda   Nguyen  Female  Alabama  31   2/17/2019   
3         Soda  ...       Linda   Nguyen  Female  Alabama  31   2/17/2019   
4         Soda  ...       Linda   Nguyen  Female  Alabama  31   2/17/2019   

  n_dependants  marital_status income  _merge  
0            3         married  40423    both  
1            3         married  40423    both  
2            3         married  40423    both  
3            3         married  40423    both  
4            3         married  40423    both  

[5 rows x 33 columns]